okio/jvm/src/main/java/okio/Pipe.kt

```diff
@@ -66,25 +74,81 @@ class Pipe(internal val maxBufferSize: Long) {
           (buffer as Object).notifyAll() // Notify the source that it can resume reading.
         }
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { write(source, byteCount) } }
     }
 
     override fun flush() {
+      var delegate: Sink? = null
       synchronized(buffer) {
         check(!sinkClosed) { "closed" }
-        if (sourceClosed && buffer.size > 0L) throw IOException("source is closed")
+
+        foldedSink?.let {
+          delegate = it
+          return@synchronized
+        }
+
+        if (sourceClosed && buffer.size > 0L) {
+          throw IOException("source is closed")
+        }
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { flush() } }
     }
 
     override fun close() {
+      var delegate: Sink? = null
       synchronized(buffer) {
         if (sinkClosed) return
+
+        foldedSink?.let {
+          delegate = it
+          return@synchronized
+        }
+
         if (sourceClosed && buffer.size > 0L) throw IOException("source is closed")
         sinkClosed = true
         (buffer as Object).notifyAll() // Notify the source that no more bytes are coming.
       }
+
+      delegate?.let { honorsTimeoutAndDeadline(it) { close() } }
     }
 
     override fun timeout(): Timeout = timeout
+
+    private fun honorsTimeoutAndDeadline(delegateSink: Sink, block: Sink.() -> Unit) {
+      val sinkTimeoutNanos = this.timeout().timeoutNanos()
+      val delegateOriginalTimeoutNanos = delegateSink.timeout().timeoutNanos()
+      val minTimeoutNanos = when {
```

Comment:
I think you want a method, `minTimeout(aNanos: Long, bNanos: Long)` that does this when block